osquery/events/linux/syslog.h

```diff
@@ -0,0 +1,172 @@
+ private:
+  friend class SyslogEventPublisher;
+};
+
+/**
+ * @brief Event details for SyslogEventPublisher events
+ */
+struct SyslogEventContext : public EventContext {
+  /**
+   * @brief The syslog message tokenized into fields.
+   *
+   * Fields will be stripped of extra space
+   */
+  std::map<std::string, std::string> fields;
+};
+
+using SyslogEventContextRef = std::shared_ptr<SyslogEventContext>;
+using SyslogSubscriptionContextRef = std::shared_ptr<SyslogSubscriptionContext>;
+
+/**
+ * @brief Event publisher for syslog lines forwarded through rsyslog
+ *
+ * This event publisher ingests JSON representations of syslog entries, and
+ * publishes them to it's subscribers. In order for it to function properly,
+ * rsyslog must be configured to forward JSON to a named pipe that this
+ * publisher will read from.
+ */
+class SyslogEventPublisher
+    : public EventPublisher<SyslogSubscriptionContext, SyslogEventContext> {
+  DECLARE_PUBLISHER("syslog");
+
+ public:
+  Status setUp() override;
+
+  void configure() override;
+
+  void tearDown() override;
+
+  Status run() override;
+
+ public:
+  SyslogEventPublisher() : EventPublisher(){};
+
+ private:
+  /// Apply normal subscription to event matching logic.
+  bool shouldFire(const SyslogSubscriptionContextRef& mc,
+                  const SyslogEventContextRef& ec) const override;
+
+  /**
+   * @brief Populate the SyslogEventContext with the syslog JSON.
+   *
+   * Performs basic cleanup on the JSON data as it is populated into the
+   * context.
+   */
+  Status populateEventContext(const std::string& line,
+                              SyslogEventContextRef& ec);
+  /**
+   * @brief Create the named pipe for log forwarding.
+   *
+   * Attempts to properly set the permissions so that rsyslog will be able to
+   * write logs to the pipe. If osquery is not running with the appropriate
+   * permissions, the named pipe permissions may have to be manually edited by
+   * the user in order for rsyslog to be able to write to it.
+   */
+  Status createFifo(const std::string& path);
+
+  time_t parseTimeString(const std::string& time_str);
+
+  /**
+   * @brief Input stream for reading from the pipe.
+   */
+  std::ifstream readStream_;
+
+ private:
+  FRIEND_TEST(SyslogTests, test_populate_event_context);
+};
+
+/**
+ * Boost TokenizerFunction functor for tokenizing rsyslog CSV data
+ *
+ * This functor is intended to be used with boost::tokenizer in order to
+ * properly parse CSV data generated by rsyslog. The default
+ * boost::escaped_list_separator provided with boost::tokenizer chokes on
+ * rsyslog CSV output, because rsyslog escapes " with "", and also does not
+ * escape backslashes. Our implementation closely follows the one provided with
+ * Boost, but allows for the idiosyncrasies of rsyslog output and simplifies
+ * the implementation for our limited use case.
+ */
+class RsyslogCsvSeparator {
+ public:
+ RsyslogCsvSeparator() : last_(false) {}
```

Comment:
Oh, I see. When I manually add the space, `make format` doesn't change it back, so at least it isn't changing to the wrong thing.